pkg/services/ngalert/state/state.go

```diff
@@ -82,11 +82,62 @@ func (a *State) GetAlertInstanceKey() (models.AlertInstanceKey, error) {
 	return models.AlertInstanceKey{RuleOrgID: a.OrgID, RuleUID: a.AlertRuleUID, LabelsHash: labelsHash}, nil
 }
 
+// Alerting sets the state to Alerting. It changes both the start and end time.
+func (a *State) Alerting(startsAt, endsAt time.Time) {
+	a.State = eval.Alerting
+	a.StateReason = ""
+	a.StartsAt = startsAt
+	a.EndsAt = endsAt
+	a.Error = nil
+}
+
+// Pending sets the state to Pending. It changes both the start and end time.
+func (a *State) Pending(startsAt, endsAt time.Time) {
+	a.State = eval.Pending
+	a.StateReason = ""
+	a.StartsAt = startsAt
+	a.EndsAt = endsAt
+	a.Error = nil
+}
+
+// NoData sets the state to NoData. It changes both the start and end time.
+func (a *State) NoData(startsAt, endsAt time.Time) {
+	a.State = eval.NoData
+	a.StateReason = ""
+	a.StartsAt = startsAt
+	a.EndsAt = endsAt
+	a.Error = nil
+}
+
+// SetError sets the state to Error. It changes both the start and end time.
+func (a *State) SetError(err error, startsAt, endsAt time.Time) {
+	a.State = eval.Error
+	a.StateReason = ""
+	a.StartsAt = startsAt
+	a.EndsAt = endsAt
+	a.Error = err
+}
+
+// Normal sets the state to Normal. It changes both the start and end time.
+func (a *State) Normal(startsAt, endsAt time.Time) {
+	a.State = eval.Normal
+	a.StateReason = ""
```

Comment:
I think this field should be an argument

Reply:
State reason is a free-form field. I think we can add as much information to it as we want.